jina/types/arrays/memmap.py

```diff
@@ -266,3 +269,72 @@ def physical_size(self) -> int:
         :return: the number of bytes
         """
         return os.stat(self._header_path).st_size + os.stat(self._body_path).st_size
+
+    def get_attributes(self, *fields: str) -> Union[List, List[List]]:
+        """Return all nonempty values of the fields from all docs this array contains
+
+        :param fields: Variable length argument with the name of the fields to extract
+        :return: Returns a list of the values for these fields.
+            When `fields` has multiple values, then it returns a list of list.
+        """
+        fields = list(fields)
+        if 'embedding' in fields and not self._embeddings_memmap:
+            embeddings = self.embeddings_memmap.tolist()
+            index = fields.index('embedding')
+            fields.remove('embedding')
+        elif 'embedding' in fields and self._embeddings_memmap:
+            embeddings = [doc.get_attributes('embedding') for doc in self]
+            self.embeddings_memmap = np.asarray(embeddings)
+            index = fields.index('embedding')
+            fields.remove('embedding')
+        if fields:
+            contents = [doc.get_attributes(*fields) for doc in self]
+            if len(fields) > 1:
+                contents = list(map(list, zip(*contents)))
+            if index:
+                contents.insert(index, embeddings)
+
+        return contents
+
+    def get_attributes_with_docs(
+        self,
+        *fields: str,
+    ) -> Tuple[Union[List, List[List]], 'DocumentArray']:
+        """Return all nonempty values of the fields together with their nonempty docs
+
+        :param fields: Variable length argument with the name of the fields to extract
+        :return: Returns a tuple. The first element is  a list of the values for these fields.
+            When `fields` has multiple values, then it returns a list of list. The second element is the non-empty docs.
+        """
+
+        contents = []
+        docs_pts = []
+
+        for doc in self:
+            contents.append(doc.get_attributes(*fields))
+            docs_pts.append(doc)
+
+        if len(fields) > 1:
+            contents = list(map(list, zip(*contents)))
+
+        return contents, DocumentArray(docs_pts)
+
+    @property
+    def _embeddings_memmap(self):
+        """Return the cached embedding stored in np.memmap.
+
+        :returns: Embeddings as np.ndarray stored in memmap.
+        """
+        return np.memmap('embedding.bin', mode='r')
+
+    @_embeddings_memmap.setter
+    def _embeddings_memmap(self, other_embeddings):
+        """Set the cached embedding values in case it is not cached.
+
+        :param other_embeddings: The embedding to be stored into numpy.memmap.
+        """
+        fp = np.memmap(
+            'embedding.bin', dtype='float32', mode='w+', shape=other_embeddings.shape
```

Comment:
LGTM, just wondering about `embedding.bin`, this seems like a relative path at the moment?  Curious if there are any plans to change it to an absolute path that's specific to each `DocumentArrayMemmap` object going forward?

Reply:
as far as know we don't have any plans yet.

Reply:
Ah okay, thanks